File: 0719-find-k-th-smallest-pair-distance/0719-find-k-th-smallest-pair-distance.cpp
class Solution {
private:
    vector<int> nums;
    int k;
    
public:
    int smallestDistancePair(vector<int>& nums, int k) {
        const int n = nums.size();
        this->nums = nums;
        this->k = k;
        
        sort(nums.begin(), nums.end());
        
        int left = 0;
        int right = nums[n - 1] - nums[0];
        int ans = 0;
        
        while (left <= right) {
            int mid = (left + right) / 2;
            int cnt = 0;
            
            for (int i = 0, j = 0; i < n; ++i) {
                while (j < n && nums[i] - nums[j] > mid) {
                    j++;
                }
                cnt += i - j; 
            }
            
            cout << mid << ' ' << cnt << '\n';
            
            if (cnt >= k) {
                ans = mid;
                right = mid - 1;
            }
            else {
                left = mid + 1;
            }
        }
        
        return ans;
    }
};

File: 0719-find-k-th-smallest-pair-distance/0719-find-k-th-smallest-pair-distance.py
class Solution:
    def smallestDistancePair(self, nums: List[int], k: int) -> int:
        nums.sort()
        n = len(nums)
        left, right = 0, 10**9
        ans = right
        
        def get_len(diff):
            glen = 0
            j = 0
            
            for i in range(n):
                while nums[i] - nums[j] > diff:
                    j += 1
                glen += i - j
            
            return glen
                
        
        while left <= right:
            mid = (left + right) // 2
            
            if get_len(mid) < k:
                left = mid + 1
            else:
                ans = mid
                right = mid - 1
                
        return ans